Gawron/ZPI-42-ZPI-64-BE-39-BE-40-Reservation-structure

Gawron97 opened the pull request:
Struktura dla rezerwacji opieki i encji zapisującej fakt samej opieki.

Gawron97 committed:
add Care entity

Gawron97 committed:
mock data generating for care

lukcyn commented on src/main/java/com/example/petbuddybackend/entity/care/Care.java:
Pododawałbym długości stringów i nullable=false na wymaganych polach

Gawron97 replied:
Znowu zapomniałem:)

lukcyn commented on src/main/java/com/example/petbuddybackend/entity/care/Care.java:
```diff
@@ -0,0 +1,65 @@
+    @Id @GeneratedValue(strategy = GenerationType.IDENTITY)
+    private Long id;
+
+    private ZonedDateTime submittedAt;
```
Rozumiem że to data utworzenia opieki? To może by dodać coś w stylu

```java
    @PrePersist
    public void prePersist() {
        submittedAt = ZonedDateTime.now();
    }
```

Gawron97 replied:
Też zapomniałem:)

Gawron97 replied:
A to działa tak że jeśli ustawie ręcznie to zostanie to co ustawiłem ręcznie, czy nadpisze?

lukcyn replied:
Tak, przed zapisaniem encji w bazie to się nadpisze

lukcyn commented on src/main/java/com/example/petbuddybackend/entity/care/Care.java:
```diff
@@ -0,0 +1,65 @@
+
+    private Double dailyPrice;
+
+    @ManyToOne(fetch = FetchType.LAZY)
+    @JoinColumn(name = "animalType", nullable = false, updatable = false)
+    private Animal animal;
+
+    @ManyToMany(fetch = FetchType.LAZY)
+    @JoinTable(
+            name = "care_animal_attribute",
+            joinColumns = @JoinColumn(name = "careId"),
+            inverseJoinColumns = @JoinColumn(name = "animalAttributeId")
+    )
+    private Set<AnimalAttribute> animalAttributes;
```
Rozumiem że animalAttribute musi dotyczyć tego animal wyżej? Jeśli tak, to dałybm jakiś check np. na PrePersist i PreUpdate źeby zapewnić że AnimalAttribute dotyczą danego Animal

Gawron97 replied:
Tylko ze nwm czy da się robić check między tabelami, trzeba go było triggera robić, chyba że @prePersistem. Sprobuje

Gawron97 replied:
W trakcie dodawania encji Care i dodania AnimalAttributes bede wyszukiwal w bazie po attributeName, AttributeValue ale tez po Animal, ktory jest przypisany do Care, wiec znajdzie mi tylko animalAttribute dla zwierzecia z Care, inaczej rzuci blad. Ale sprobuje dodac tego PrePersista

lukcyn commented on src/main/java/com/example/petbuddybackend/entity/care/Care.java:
```diff
@@ -0,0 +1,65 @@
+@Builder
+@Entity
+@Getter @Setter
+@Table(
+        uniqueConstraints = { @UniqueConstraint(columnNames = { "caretakerEmail", "clientEmail", "careStart", "careEnd" }) }
+)
```
Czy dałoby radę z tą konfiguracją dodać np. dwa psy, pare różnych zwierząt w tym samym czasie i u tej samej osoby?

Gawron97 replied:
Hmm, rzeczywiście nie dałoby się. Nie wiem czy unique tutaj wgl ma sens w przypadku tej encji

lukcyn commented on src/main/java/com/example/petbuddybackend/entity/care/CareStatus.java:
```diff
@@ -0,0 +1,12 @@
+package com.example.petbuddybackend.entity.care;
+
+public enum CareStatus {
+
+    PENDING,
+    ACCEPTED,
+    REJECTED,
+    CANCELLED,
+    OUTDATED,
+    PAID
+
+}
```
- W sumie czym się cancelled od rejected różni? 
- Może by dodać AWAITING_PAYMENT żeby dać znać obu stronom że trzeba przelew zrobić?
- outdated to oznacza że usługa zakończona? Bo jeśli tak to dałbym COMPLETED

Gawron97 replied:
Outdated to znaczy że nikt nie zatwierdził a data obecną już jest po dacie początku opieki(albo dzien przed, do ustalenia)

Gawron97 replied:
Awaiting paymeny można dać
Rejected to że opiekun ksrzucil a cancelled to że klient jednak się wycofał, chyba bardziej pasuje ten rozdział do obu osobno, ale można ujednolicic

lukcyn commented on src/main/java/com/example/petbuddybackend/entity/care/Care.java:
```diff
@@ -0,0 +1,65 @@
+
+    private String description;
+
+    private Double dailyPrice;
```
Może troche późno, ale doczytałem że do walut lepszy jest BigDecimal. Przez precyzję Double i Float mogą być problemy z tym w przyszłości

Gawron97 replied:
Aa racja

Gawron97 committed:
change prices to BigDecimal
remove unique constraints of Care entity
add columns constraints for Care entity

lukcyn reviewed (approved):
Git

Gawron97 committed:
add @PrePersist to check if animal attributes match with animal provided in offer or in care
